Use Gradle's variant-aware dependency management and allow using Gradle's built-in Java Module System support

jjohannes opened the pull request:
Hi 👋 I recently became aware of the current state of JavaFX support in Gradle and there are two things that are quite painful for some users right now. Both could be improved, if the implementation in this plugin (and maybe later JavaFX itself) would rely on Gradle's [variant-aware dependency management](https://docs.gradle.org/current/userguide/variant_model.html) and [Gradle Metadata](https://docs.gradle.org/current/userguide/publishing_gradle_module_metadata.html). 

## The issues
1. It's hard to publish libraries relying on JavaFX with good metadata due to the cross-cutting concern of selecting the right "JavaFX Platform Jars". (Which is one of the main use cases that variant-aware dependency management with variant selection based on attributes solves.) See e.g.: https://github.com/gradle/gradle/issues/26144
2. This plugin is bound to  another plugin –`org.javamodularity.moduleplugin` – right now. That does not work well together with Gradle's built-in Module System support (added in Gradle 6.4). It would be good if users of JavaFX can choose which of the mechanism they like to use.

## Details

(My background is that I worked on both the dependency management and the Java Module support in Gradle – I was part of the Gradle core development team for several years – and continue contributing in these areas.)

This PR proposes a full implementation. It also extends the test coverage to show that things still work as before (from the user perspective) but better. It purposefully does not change the API of `JavaFXOptions` so that the changes should not have an effect on how users use this plugin. Before the actual code change, this PR contains three commits for the following:
1. Updating the Gradle Wrapper to the current release (8.3)
2. Updating the "build.gradle" and "settings.gradle" and the plugins used in them to the latest standards and features. Basically doing some cleanup there. I followed what we do for all the plugins of https://gradlex.org/
3. Updating and extending the smoke test. It now:
     - Tests against multiple Gradle versions
     - Checks the actual `-classpath` and `--module-path` parameters used when Gradle calls `javac` and `java` by parsing Gradle's `--debug` output.
     - Has two additional tests: building with a local JavaFX sdk, using Gradle's built-in Module System support (since Gradle 6.4)

## Main code changes
- The JavaFX platform selection is now based on _variants_. This mechanism is available since Gradle 4/5 and stable since Gradle 6. It's prominently used for instance in [Kotlin Multi-Platform](https://kotlinlang.org/docs/multiplatform-publish-lib.html) where many libraries have similar requirements (different artifacts for different target platforms). Here are some docs. I am happy to explain more if there are questions:
    - Docs: https://docs.gradle.org/current/userguide/variant_model.html
    - Example with native Jars: https://docs.gradle.org/current/userguide/component_metadata_rules.html#adding_variants_for_native_jars
    - Discussions: https://github.com/gradlex-org/java-ecosystem-capabilities/issues/57, https://github.com/gradle/gradle/issues/26144#issuecomment-1686037462
- Then, based on that or as general improvement/simplification
     - Transitive dependencies no longer need to be maintained in the plugin
     - Empty Jars are completely gone from the classpath autoamtically (see test assertions)
     - Manual classpath and module path modification limited to minimum (only required for running non-modular applications, as Gradle does not support non-modular Apps with some Jars on the --module-path yet directly)
     - No hard dependency on 'org.javamodularity:moduleplugin' anymore
     - Dependency declaration done lazily (via withDependencies)


With these changes, the minimal supported Gradle version becomes `6.1` as only then all the dependency management APIs used are available.

## Next steps / JavaFX and Gradle Metadata

The change contains a `JavaFXComponentMetadataRule`. This extends the metadata of JavaFX. In the future, JavaFX could publish this information (using Gradle Metadata) directly. Then this plugin will mostly become a convenience for defining dependencies to JavaFX. But, in particular in Module Path-based projects, JavaFX could even be used without this plugin.

-----

I know this is a rather large change. I can add more documentation and explain things in more detail. I wanted to push this out to get the discussion started.

samypr100 commented:
FWIW I can at least confirm that this is working on a few projects that I quickly tried it ([FXyz](https://github.com/FXyz/FXyz), [litfx](https://github.com/Birdasaur/LitFX), [controlsfx](https://github.com/controlsfx/controlsfx), [hansolo charts](https://github.com/HanSolo/charts), [trinity](https://github.com/Birdasaur/Trinity), etc.) on. It generates the GMM and POM I would expect.

Was able to even remove the manual pom modifiers used on some of these projects to remove the classifier from the generated maven poms.

```groovy
pom.withXml {
    asNode().dependencies.'*'
            .findAll() { it.groupId.text() == 'org.openjfx' }
            .each { it.remove(it.classifier) }
}
```

One thing worthy of mention is that projects that relied knowingly/unknowingly on the transitive 'org.javamodularity.moduleplugin' will now have to be explicit about it otherwise their builds could likely fail as a result, hence I’d consider that a breaking change from a plugin perspective so maybe it deserves a bigger version bump, or doc updates highlighting this change.

johanvos commented:
Thanks for the PR, I believe that can indeed be helpful to developers.
Pinging @abhinayagarwal , @jperedadnr  or @tiainen : can one of you review this?

jperedadnr commented on src/test/java/org/openjfx/gradle/JavaFXPluginSmokeTestGradle61.java:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (c) 2018, 2023, Gluon
```
copyright header of new files: 
```
Copyright (c) 2023, Gluon
```

jperedadnr commented on src/test/java/org/openjfx/gradle/JavaFXPluginSmokeTestGradle76.java:
```diff
@@ -0,0 +1,38 @@
+ */
+package org.openjfx.gradle;
+
+class JavaFXPluginSmokeTestGradle76 extends JavaFXPluginSmokeTest{
```
minor, add blank space before `{`

jperedadnr commented on src/test/java/org/openjfx/gradle/JavaFXPluginSmokeTestGradle83.java:
```diff
@@ -0,0 +1,38 @@
+ */
+package org.openjfx.gradle;
+
+class JavaFXPluginSmokeTestGradle83 extends JavaFXPluginSmokeTest{
```
minor, add blank space before `{`

jperedadnr commented on test-project/local-sdk/build.gradle:
```diff
@@ -0,0 +1,17 @@
+}
+
+javafx {
+    modules = ['javafx.controls', 'javafx.base', 'javafx.graphics']
```
base and graphics are not needed there (derived from controls)

jjohannes replied:
That won't work, because here we do not have the dependencies from the metadata (dependencies from POM files). The Jar files will be missing on the module path.

But this probably worked on _master_, now that I think about it. Maybe it should be changed then. I can think of the following options:

1. Accept it like this
2. Add back the dependency information to `JavaFXModule` and use it only for the "local-sdk" case
3. Put all Jars on the module path (for this 'local-sdk' case) no matter what is declared. Java will take care of only loading what is needed and as the Jars are already there, Gradle does not need to download them. 

What do you think?

jperedadnr replied:
On the Maven side it happens somehow the same: when using a local SDK we also need to add all the JavaFX dependencies:

```
         <profile>
            <id>sdk</id>
            <dependencies>
                <dependency>
                    <groupId>org.openjfx</groupId>
                    <artifactId>javafx-base</artifactId>
                    <version>${javafx.version}</version>
                    <scope>system</scope>
                    <systemPath>${javafx.systemPath}/javafx.base.jar</systemPath>
                </dependency>
...
```

So the question here is: for anyone using `sdk` (not only for tests), `modules` will have to change to specify _all_ the JavaFX modules after this PR, while that was working before it? 

This is definitely a breaking change, that could be admitted, but will need some documentation. 

Else, if we keep JavaFXModule as it was, then nothing will change. Imo, keeping that dependency information is not a big issue.

The option of adding all jars found in the sdk path to the module path seems a little bit undefined, as that path might contain all or just a few jars or even unrelated jars.

jjohannes replied:
>  if we keep JavaFXModule as it was, then nothing will change. Imo, keeping that dependency information is not a big issue.

I added the information back and only use it in the "Local SDK" case. I added a comment to explain this special case. This way it works as before when using a local SDK.

jperedadnr commented on test-project/local-sdk/javafx-sdk-17.0.8/lib/javafx-swt.jar:
No needed for the local-sdk test (only controls, graphics and base)

jperedadnr commented on test-project/modular-with-modularity-plugin/src/main/java/org/openjfx/gradle/javafx/test/Main.java:
```diff
@@ -0,0 +1,30 @@
+package org.openjfx.gradle.javafx.test;
```
missing license header

jperedadnr commented on test-project/modular-with-modularity-plugin/src/main/java/module-info.java:
```diff
@@ -0,0 +1,5 @@
+module org.openjfx.gradle.javafx.test {
```
missing license header

jperedadnr commented on test-project/local-sdk/src/main/java/org/openjfx/gradle/javafx/test/Main.java:
```diff
@@ -0,0 +1,30 @@
+package org.openjfx.gradle.javafx.test;
```
missing license header

jperedadnr commented on test-project/transitive/build.gradle:
```diff
@@ -1,11 +1,29 @@
 javafx {
     version = '17'
-    modules = [ 'javafx.controls' ]
+    modules = [ 'javafx.controls', 'javafx.media', "javafx.swing", "javafx.web" ]
```
web includes controls and media, but why do you add swing/web to this test?

jjohannes replied:
There were version inconsistencies, because the transitively pulled in the "media, swing, web" dependencies on `11`. Declaring them here also updated them to 17.

It was already a problem on _master_ I believe, but there the test only checked that the transitive dependencies did not break the build. But the result was actually missing the platform Jars for the three transitive dependencies - media, swing, web - and only contained the empty Jars for these. This problem is now fixed.

I changed the test to use another example and have transitive dependencies on `17` versions pulled in. I think that's a better test and I reverted the change to this line.

jperedadnr reviewed (commented):
I've done a quick review, it seems it works fine (tests work, and updated HelloFX samples work)
I have some somehow minor comments, and will give another review once these are addressed.

jjohannes commented on src/main/java/org/openjfx/gradle/JavaFXOptions.java:
```diff
@@ -130,64 +174,52 @@ public List<String> getModules() {
                     .forEach(javaFXModule -> {
                         if (customSDKArtifactRepository != null) {
-                            project.getDependencies().add(conf, Map.of("name", javaFXModule.getModuleName()));
+                            dependencySet.add(getDependencies().create(Collections.singletonMap("name", javaFXModule.getModuleName())));
```
Sorry this was unintended. Reverted. 
Thanks for checking so thoroughly.

jjohannes commented on src/main/java/org/openjfx/gradle/JavaFXModule.java:
```diff
@@ -43,22 +42,16 @@
-    MEDIA(BASE, GRAPHICS),
-    SWING(BASE, GRAPHICS),
-    WEB(BASE, CONTROLS, GRAPHICS, MEDIA);
+    GRAPHICS,
```
Yes. The JavaFX platform Jar selection with classifiers did not work before in Gradle for transitive dependencies. You just got the empty (unclassified) Jars. The variant selection based on attributes selects variants across the whole dependency graph.

Basically the plugin now does this:
```
        classpath.getAttributes().attribute(
                OperatingSystemFamily.OPERATING_SYSTEM_ATTRIBUTE,
                getObjects().named(OperatingSystemFamily.class, platform.getOsFamily()));
        classpath.getAttributes().attribute(
                MachineArchitecture.ARCHITECTURE_ATTRIBUTE,
                getObjects().named(MachineArchitecture.class, platform.getArch()));
```

Which influences the selection for each component in the dependency graph, no matter how deep it is inside.

Before the plugin declared dependencies with classifiers, which only influenced these directly declared dependencies.

jjohannes commented:
Thank you for the quick response and the thorough review @jperedadnr. I addressed the feedback and answered the questions. Let me know if you have more.

jperedadnr commented on src/main/java/org/openjfx/gradle/JavaFXPlugin.java:
```diff
@@ -29,21 +29,100 @@
+        ));
+    }
 
-        project.getTasks().create("configJavafxRun", ExecTask.class, project);
```
This was removed (and the `ExecTask` implementation).

One of its purposes (maybe the only one at this point) was supporting the `nativeRunAgent` task from the GluonFX gradle plugin.

See: https://github.com/gluonhq/gluonfx-gradle-plugin/blob/master/src/main/java/com/gluonhq/gradle/tasks/NativeRunAgentTask.java#L60 and following lines.

Main purpose of this is: 
- JavaFX plugin: `./gradlew run` -> runs `run` task, with current JAVA_HOME
- GluonFX plugin: `./gradlew nativeRunAgent` -> runs JavaFX plugin's `run` task, but with GRAALVM_HOME and adds the native agent arguments

Probably there is a better (Gradle) way to do it, but as is now, I believe the GluonFX plugin can be updated to not require this task anymore.

Thoughts?

jjohannes replied:
Configuring one task by first calling another task that does the configuration is not how things should be done in Gradle. And most (re)configuration done there is no longer needed. That's why I removed it.

Ideally all configuration is just done in the normal configurations code without doing anything at "task execution time". Sometimes if you need to do something "hacky", because a certain thing is not well supported, you can do it in a `doFirst {}` action of the task you need to modify. The changes in this PR do that for modifying the `--modue-path` in the case of the NON-Modular application.

I think in the _GluonFX gradle plugin_ you can now remove these lines:
https://github.com/gluonhq/gluonfx-gradle-plugin/blob/master/src/main/java/com/gluonhq/gradle/tasks/NativeRunAgentTask.java#L79-L83

```
Task javafxRun = project.getTasks().findByName(CONFIG_JAVAFX_RUN_TASK);
        if (javafxRun == null) {
            throw new GradleException("javafxplugin:" + CONFIG_JAVAFX_RUN_TASK + " task not found.");
        }
        this.dependsOn(javafxRun.getPath());
```

As the configuration this task did is no longer necessary.

Later in the  _GluonFX gradle plugin_ you look at the "run" task, which will have everything configured, except for what we now do in the `doFirst {}` action. But you do that part already yourself in the GluonFX plugin:
https://github.com/gluonhq/gluonfx-gradle-plugin/blob/master/src/main/java/com/gluonhq/gradle/tasks/NativeRunAgentTask.java#L126-L131

Would it be okay for you to remove
https://github.com/gluonhq/gluonfx-gradle-plugin/blob/master/src/main/java/com/gluonhq/gradle/tasks/NativeRunAgentTask.java#L79-L83
and update that plugin as well?

Maybe you can test these changes together to confirm that everything still works.

jperedadnr replied:
Yes, I've already tried that. Removing those lines worked fine, I could run it without issues. 
One we move forward with this PR, we'll have to revisit the GluonFX plugin.

jjohannes replied:
What we could do is register a no-op task with the name that we don't use. Then the other plugin should continue to work as it is. And you can remove the no-op in a later cleanup here.

`project.gerTasks().register("configJavafxRun")`

jperedadnr reviewed (approved):
Looks good, I've done some testing with different JavaFX Gradle projects, all look good (after some updates that are needed once this PR is merged and a new version of the plugin is published).

abhinayagarwal reviewed (approved):
LGTM. Tested both modular and non-modular applications and libraries.

jjohannes commented on src/main/java/org/openjfx/gradle/JavaFXOptions.java:
```diff
@@ -130,64 +174,58 @@ public List<String> getModules() {
-        lastUpdatedConfigurations = configurations;
+            var javaFXModules = JavaFXModule.getJavaFXModules(this.modules);
+            if (customSDKArtifactRepository == null) {
+                javaFXModules.stream().sorted().forEach(javaFXModule -> dependencySet.add(getDependencies().create(
```
Sure. I have no strong opinion on that. I reformatted, using what your proposal.

jperedadnr reviewed (approved):
Looks good!

nlisker commented:
I think that this will fix the following issues:
https://github.com/openjfx/javafx-gradle-plugin/issues/133
https://github.com/openjfx/javafx-gradle-plugin/issues/126
https://github.com/openjfx/javafx-gradle-plugin/issues/94

Possibly https://github.com/openjfx/javafx-gradle-plugin/issues/26 too.

jperedadnr commented:
Okay, let's merge and test before doing a new release, and see if those issues can be closed.

nlisker commented:
I can test on my local pet projects where I hit this bug. If there's a nightly/integration version out I will test before a stable release is made.

jperedadnr commented:
The snapshot gets published to Sonatype, you just need:

```
pluginManagement {
    repositories {
        maven {
            url = "https://oss.sonatype.org/content/repositories/snapshots"
        }
        gradlePluginPortal()
    }
}
```

nlisker commented:
I tested `15-SNAPSHOT` and can confirm that, in Eclipse, the JavaFX modules now appear as "is modular" while "Project and External Dependencies" is on the classpath, and the test libraries (JUnit) appear as "not modular". This allows the tests to work on the classpath as they should, while there are no JavaFX missing modules in `module-info.java`.

![image](https://github.com/openjfx/javafx-gradle-plugin/assets/37422899/2e570f9a-52e1-4211-b4b7-128ed25086ae)

I also rolled back to version 14 and saw the problem reappearing, so the snapshot version definitely saves the day here.

Thanks @jjohannes!

nlisker commented:
I'll note that this revives a problem with the gluonfx gradle plugin that was partially fixed in https://github.com/gluonhq/gluonfx-gradle-plugin/pull/107. I can provide a reproduceable example on that project's repo if needed.

abhinayagarwal commented:
@nlisker can you please open a new issue for this?

nlisker commented:
> @nlisker can you please open a new issue for this?

Created https://github.com/gluonhq/gluonfx-gradle-plugin/issues/179
